feat: cache toolset status and add toolset management tool command

mainred opened the pull request:
- use isort to sort the imported packaged
- introduce toolset_manager to manage toolsets, which was Config's job.
- only enable build-in toolsets specified in the config file
- use load_toolsets_config as the only entrypoint to load initialize and validate a toolset from a config/definition
- introduce cli command to list and refresh the toolset status from local cache

```
 ./dist/holmes/holmes toolset --help

 Usage: holmes toolset [OPTIONS] COMMAND [ARGS]...

 toolset management commands

╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --help          Show this message and exit.                                                                                                           │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ list      List build-int and custom toolsets status of CLI                                                                                            │
│ refresh   Refresh build-in and custom toolsets status of CLI                                                                                          │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

```

```
./dist/holmes/holmes toolset list
                                                   Toolset Status
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃ Name                             ┃ Status   ┃ Enabled ┃ Type       ┃ Path                                ┃ Error ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│ confluence                       │ disabled │ False   │ built-in   │ None                                │ None  │
│ docker/core                      │ disabled │ False   │ built-in   │ None                                │ None  │
│ helm/core                        │ disabled │ False   │ built-in   │ None                                │ None  │
│ kubernetes/logs                  │ disabled │ False   │ built-in   │ None                                │ None  │
│ servicenow                       │ disabled │ False   │ built-in   │ None                                │ None  │
│ aks/node-health                  │ disabled │ False   │ built-in   │ None                                │ None  │
│ aws/security                     │ disabled │ False   │ built-in   │ None                                │ None  │
│ aws/rds                          │ disabled │ False   │ built-in   │ None                                │ None  │
│ argocd/core                      │ disabled │ False   │ built-in   │ None                                │ None  │
│ slab                             │ disabled │ False   │ built-in   │ None                                │ None  │
│ aks/core                         │ disabled │ False   │ built-in   │ None                                │ None  │
│ kubernetes/core                  │ enabled  │ True    │ built-in   │ None                                │ None  │
│ kubernetes/live-metrics          │ disabled │ False   │ built-in   │ None                                │ None  │
│ kubernetes/kube-prometheus-stack │ disabled │ False   │ built-in   │ None                                │ None  │
│ kubernetes/krew-extras           │ disabled │ False   │ built-in   │ None                                │ None  │
│ kubernetes/kube-lineage-extras   │ disabled │ False   │ built-in   │ None                                │ None  │
│ internet                         │ disabled │ False   │ built-in   │ None                                │ None  │
│ robusta                          │ disabled │ False   │ built-in   │ None                                │ None  │
│ opensearch/status                │ disabled │ False   │ built-in   │ None                                │ None  │
│ grafana/loki                     │ disabled │ False   │ built-in   │ None                                │ None  │
│ grafana/tempo                    │ disabled │ False   │ built-in   │ None                                │ None  │
│ newrelic                         │ disabled │ False   │ built-in   │ None                                │ None  │
│ grafana/grafana                  │ disabled │ False   │ built-in   │ None                                │ None  │
│ notion                           │ disabled │ False   │ built-in   │ None                                │ None  │
│ kafka/admin                      │ disabled │ False   │ built-in   │ None                                │ None  │
│ datadog                          │ disabled │ False   │ built-in   │ None                                │ None  │
│ prometheus/metrics               │ disabled │ False   │ built-in   │ None                                │ None  │
│ datetime                         │ enabled  │ True    │ built-in   │ None                                │ None  │
│ opensearch/logs                  │ disabled │ False   │ built-in   │ None                                │ None  │
│ opensearch/traces                │ disabled │ False   │ built-in   │ None                                │ None  │
│ coralogix/logs                   │ disabled │ False   │ built-in   │ None                                │ None  │
│ rabbitmq/core                    │ disabled │ False   │ built-in   │ None                                │ None  │
│ git                              │ disabled │ False   │ built-in   │ None                                │ None  │
│  ig/core                          │ enabled  │ True    │ customized │ /home/azureuser/inspect_gadget.yaml │ None  │
```

mainred commented on holmes/core/toolset_manager.py:
```diff
@@ -0,0 +1,314 @@
+        # custom toolset should not override built-in toolsets
+        # to test the new change of built-in toolset, we should make code change and re-compile the program
+        custom_toolsets = self.load_custom_toolsets()
+        for custom_toolset in custom_toolsets:
```
server is using custome_toolset to enable built-in toolsets and add extra customized toolset.
We need to consider add custom config toolset overriding built-in toolset back to not block server.

https://github.com/robusta-dev/holmesgpt/blob/master/helm/holmes/templates/toolset-config.yaml

mainred commented on holmes/core/toolset_manager.py:
```diff
@@ -0,0 +1,314 @@
+                toolsets=self.toolsets, strict_check=False
+            )
+            if toolsets_from_config:
+                self.add_or_merge_yaml_toolsets(
```
There's some misunderstanding and less documented part for tool_config.
From the original code, both toolset and custom_toolset can override built-in toolset and introduce new customized toolset, which seems to be redundant.

moshemorad commented on holmes/core/toolset_manager.py:
```diff
@@ -0,0 +1,394 @@
+        ]
+
+    # TODO(mainred): cache and refresh periodically toolset status for server if necessary
+    def list_enabled_server_toolsets(
```
Looks like for some reason when running with this version on my local cluster the datetime toolset isn't being enabled by default.

moshemorad commented on holmes/core/toolset_manager.py:
```diff
@@ -0,0 +1,394 @@
+
+        return all_toolsets_with_status
+
+    def list_enabled_console_toolsets(
```
The current implementation cause it to disable all toolset by default.
This is different behiour then our initial intantion. We would like that every new user will have toolsets that are enabled by default to start and get the hand of holmes before messing with toolsets config. This is quite a change and i think that we need to be aligned on that.
@arikalon1 WDYT?

mainred replied:
Thanks for sharing the intention of the design.

My intention is to disable all tools by default, and customer has to enable the tools explicitly. This allows us to reduce the chance the loading unprepared toolset.

moshemorad reviewed (changes requested):
There are few changes that effect the default user expiriance. we need to align on that.

mainred commented on holmes/core/tools.py:
```diff
@@ -345,14 +326,15 @@ class Toolset(BaseModel):
-    _status: ToolsetStatusEnum = PrivateAttr(ToolsetStatusEnum.DISABLED)
-    _error: Optional[str] = PrivateAttr(None)
+    # status fields that be cached
+    type: ToolsetType = ToolsetType.BUILTIN
```
yeah, in the code path, we always set the toolset type. I can make it optional, and let the loading toolset part set the toolset type.

mainred commented on holmes/core/tools.py:
```diff
@@ -345,14 +326,15 @@ class Toolset(BaseModel):
-    _error: Optional[str] = PrivateAttr(None)
+    # status fields that be cached
+    type: ToolsetType = ToolsetType.BUILTIN
+    path: Optional[FilePath] = None
```
Do you know how can we detect this error from any lint?

mainred commented on holmes/core/toolset_manager.py:
```diff
@@ -0,0 +1,394 @@
+        self.toolset_status_location = toolset_status_location
+
+    @property
+    def cli_tool_tags(self) -> List[str]:
```
We don't recommend using list for in checking?

mainred commented on holmes/main.py:
```diff
@@ -126,19 +139,15 @@ def init_logging(verbose_flags: Optional[List[bool]] = None):
 opt_model: Optional[str] = typer.Option(None, help="Model to use for the LLM")
 opt_config_file: Optional[Path] = typer.Option(
-    None,
+    DEFAULT_CONFIG_LOCATION,
```
I can  change the code to make it make the configurable optional

mainred replied:
It's because I used the var before it's initialized. Will create the config_from_file first

mainred commented on holmes/plugins/toolsets/__init__.py:
```diff
@@ -81,13 +77,79 @@ def load_python_toolsets(dal: Optional[SupabaseDal]) -> List[Toolset]:
+    if is_old_toolset_config(toolsets):
+        message = "Old toolset config format detected, please update to the new format: https://docs.robusta.dev/master/configuration/holmesgpt/custom_toolsets.html"
+        logging.warning(message)
+        raise ValueError(message)
```
I think it's good to remove it if the format is not supported anymore after we announce deprecating it several releases.

moshemorad commented:
> I have resolved code conflicts multiple times since my change touch the config.py. I'll focus on the PR comments if there are incoming ones, and until all looks good, I'll resolve the conflicts. cc @moshemorad

The fix is due to the missed error when we changed private attr in the Toolset Model. So it might raise more issue. Let me know if you bump into anything that need my help.

mainred commented:
> Let me know if you bump into anything that need my help.

This PR is ready for review. Please take another look. Thanks.

mainred commented:
Please ignore the conflicts lets focus on the PR change first. I will definitely resolve the conflicts after addressing all comments to reduce the chance of resolving conflicts.

mainred commented on holmes/config.py:
```diff
@@ -170,7 +96,12 @@ class Config(RobustaBaseConfig):
     custom_runbooks: List[FilePath] = []
-    custom_toolsets: List[FilePath] = []
+
+    # custom_toolsets is defined in the config file, and can override built-in toolsets as stable custom toolsets
```
re-phrase the comments for custom_toolsets and custom_toolsets_from_cli, please take a look.

mainred commented:
@moshemorad could you please help find the root cause of llm evaluation error. The following error message raised from the pipeline seems to be common issue? Thanks.
`httpx.LocalProtocolError: Illegal header value b'Bearer '

moshemorad commented:
@mainred Looks like the issue is due to forked repos don't use the original repo secrets.
https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions#using-secrets-in-a-workflow

@arikalon1 @nherment any idea what we want to do in those cases?

moshemorad commented on holmes/core/tools.py:
```diff
@@ -323,7 +301,10 @@ class ToolsetEnvironmentPrerequisite(BaseModel):
+
+    # Built-in toolset is enabled when 1. it's specified in the config and 2. 'enabled' is not set to false
+    # Custom toolset is enabled when 'enabled' is not set to false
+    enabled: bool = True
```
This change causing to scenario when we are trying to load a lot of irrelevant tools in the server and most of the in error state. 
![image](https://github.com/user-attachments/assets/37224b6f-d902-43e4-a3d4-303a0ae15766)

We need to make sure that in server it will be as today.
![image](https://github.com/user-attachments/assets/244f524c-dfb7-4b77-a844-b7baf380f738)
